### PR TITLE
core: Make PatternRewriter a Builder

### PR DESCRIPTION
**Pull request by math-fehr**
Stacked PRs:
 * __->__#3540
 * #3539
 * #3538
 * #3537

--- --- ---

### core: Make PatternRewriter a Builder

This allows us to now use the ImplicitBuilder on PatternRewriter,
or to use `insert` using an InsertPoint.

### CODE REVIEW

**Review by superlopuh — APPROVED** [review]
Fantastico

File: xdsl/pattern_rewriter.py [hunk 1]
```diff
@@ -91,6 +91,11 @@ class PatternRewriter(PatternRewriterListener):
     has_done_action: bool = field(default=False, init=False)
     """Has the rewriter done any action during the current match."""
 
+    def __init__(self, current_operation: Operation):
```
Comment (alexarice) [inline xdsl/pattern_rewriter.py:94]: You could possibly do this with a `__post_init__`

File: xdsl/pattern_rewriter.py [hunk 2]
```diff
@@ -91,6 +91,11 @@ class PatternRewriter(PatternRewriterListener):
     has_done_action: bool = field(default=False, init=False)
```
Comment (alexarice) [inline xdsl/pattern_rewriter.py:91]:
```suggestion
    has_done_action: bool
```
I think `field` is dead code if you add an init?

File: xdsl/pattern_rewriter.py [hunk 3]
```diff
@@ -726,7 +731,11 @@ def _handle_operation_removal(self, op: Operation) -> None:
         """Handle removal of an operation."""
         if self.apply_recursively:
             self._add_operands_to_worklist(op.operands)
-        self._worklist.remove(op)
+        if op.regions:
```
Comment (alexarice) [inline xdsl/pattern_rewriter.py:734]: Is this a separate change to making `PatternRewriter` a `Builder`?

File: xdsl/pattern_rewriter.py [hunk 4]
```diff
@@ -726,7 +731,11 @@ def _handle_operation_removal(self, op: Operation) -> None:
...
+            for sub_op in op.walk():
+                self._worklist.remove(sub_op)
+        else:
+            self._worklist.remove(op)
```
Comment (alexarice) [inline xdsl/pattern_rewriter.py:738]:
```suggestion
        self._worklist.remove(op)
```
Surely the operation itself still needs to be removed when it has regions?

File: xdsl/rewriter.py [hunk 5]
```diff
@@ -69,10 +69,10 @@ def erase_op(op: Operation, safe_erase: bool = True):
...
         Otherwise, replace its uses with ErasedSSAValue.
         """
-        assert op.parent is not None, "Cannot erase an operation that has no parents"
-
-        block = op.parent
-        block.erase_op(op, safe_erase=safe_erase)
+        if (block := op.parent) is not None:
+            block.erase_op(op, safe_erase=safe_erase)
+        else:
+            op.erase(safe_erase=safe_erase)
```
Comment (alexarice) [inline xdsl/rewriter.py:75]: Why is it now ok to erase a operation with no parents?

File: xdsl/transforms/convert_scf_to_openmp.py [hunk 6]
```diff
@@ -43,7 +44,8 @@ def match_and_rewrite(self, loop: scf.ParallelOp, rewriter: PatternRewriter, /):
             regions=[Region(Block())],
             operands=[[], [], [], [], [], []],
         )
-        with ImplicitBuilder(parallel.region):
+        rewriter.insertion_point = InsertPoint.at_end(parallel.region.block)
+        with ImplicitBuilder(rewriter):
```
Comment (alexarice) [inline xdsl/transforms/convert_scf_to_openmp.py:48]: I don't think I'm a fan of this API. Does the user need to reset the insertion point of the rewriter after doing this?

**Review by alexarice — COMMENTED** [review]
I left some comments but they may be more due to my ignorance than actual problems with the change.

File: xdsl/backend/riscv/riscv_scf_to_asm.py [hunk 7]
```diff
@@ -69,15 +69,15 @@ def match_and_rewrite(self, op: riscv_scf.ForOp, rewriter: PatternRewriter, /):
         yield_op = body.last_op
         assert isinstance(yield_op, riscv_scf.YieldOp)
 
-        body.insert_ops_after(
+        rewriter.insert_op(
```
Comment (compor) [inline xdsl/backend/riscv/riscv_scf_to_asm.py:72]: What is the benefit of this change?

**Review by compor — APPROVED** [review]

File: xdsl/rewriter.py (hunk 5 above)
Comment (superlopuh) [inline xdsl/rewriter.py:75]: TBH this feels like it could be its own PR, with an accompanying test

File: xdsl/pattern_rewriter.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/pattern_rewriter.py:94]: The `self.current_operation = current_operation` seems like strictly `__init__` code to me

Comment (math-fehr) [reply xdsl/pattern_rewriter.py:94]: I cannot actually do a `__post_init__`, because if I would use the generated `__init__`, this would require the arguments for `PatternRewriterListener` and `Builder`. But the `Builder` arguments gets constructed in the `__init__` I wrote here.

File: xdsl/pattern_rewriter.py (hunk 2 above)
Comment (math-fehr) [reply xdsl/pattern_rewriter.py:91]: Nope, the `default` still get added. I removed the `init=False` though, as we don't have an `__init__` here.

File: xdsl/pattern_rewriter.py (hunk 3 above)
Comment (math-fehr) [reply xdsl/pattern_rewriter.py:734]: I will rebase this branch on a new one, this was indeed a bug before that can be refactored.

File: xdsl/rewriter.py (hunk 5 above)
Comment (math-fehr) [reply xdsl/rewriter.py:75]: I'll move this to a separate PR as well, the idea is that sometimes we detach an operation and then destroy it.

File: xdsl/transforms/convert_scf_to_openmp.py (hunk 6 above)
Comment (math-fehr) [reply xdsl/transforms/convert_scf_to_openmp.py:48]: I agree the API is not optimal.
Currently, what we need to use `ImplicitBuilder` is:
* set the insertion point on the rewriter
* Create the ImplicitBuilder using the rewriter

We don't need to reset it after though.

What would you think of `ImplicitBuilder(rewriter, at=InsertPoint.at_end(...))`, which would set the insertion point just for that region of code?

File: xdsl/backend/riscv/riscv_scf_to_asm.py (hunk 7 above)
Comment (math-fehr) [reply xdsl/backend/riscv/riscv_scf_to_asm.py:72]: If we don't notify the `PatternRewriter` that these operations are added, they are not matched anymore by the `PatternRewriter`.

**Comment by math-fehr** [comment]
Since some time passed, and to simplify splitting this PR, I created a new stack of PRs starting at math-fehr/stack/5
It contains the same things but splitted on 3 different PRs, with more tests.
